apps/balancer-tools/src/app/poolsimulator/(components)/ImpactCurve.tsx

```diff
@@ -1,72 +1,101 @@
 "use client";
 
-import { AMM } from "@bleu-balancer-tools/math-poolsimulator/src";
-import { PoolPairData } from "@bleu-balancer-tools/math-poolsimulator/src/types";
 import { PlotType } from "plotly.js";
+import { useEffect, useState } from "react";
 
 import Plot from "#/components/Plot";
 import { Spinner } from "#/components/Spinner";
 import { usePoolSimulator } from "#/contexts/PoolSimulatorContext";
 import { formatNumber } from "#/utils/formatNumber";
 
-import { PoolTypeEnum, TokensData } from "../(types)";
-import { calculateCurvePoints, trimTrailingValues } from "../(utils)";
+import { PoolTypeEnum } from "../(types)";
+import {
+  ImpactWorkerInputData,
+  ImpactWorkerOutputData,
+} from "../(workers)/impact-curve-calculation";
+
+interface Amounts {
+  analysisTokenIn?: {
+    amounts: number[];
+    priceImpact: number[];
+  };
+  tabTokenIn?: {
+    amounts: number[];
+    priceImpact: number[];
+  };
+}
 
 export function ImpactCurve() {
-  const {
-    analysisToken,
-    currentTabToken,
-    initialAMM,
-    customAMM,
-    initialData,
-    customData,
-  } = usePoolSimulator();
+  const { analysisToken, currentTabToken, initialData, customData } =
+    usePoolSimulator();
+
+  const [initialAmounts, setInitialAmounts] = useState<Amounts>({});
+  const [customAmounts, setCustomAmounts] = useState<Amounts>({});
+
+  if (!initialData || !customData) return <Spinner />;
 
-  if (!initialAMM || !customAMM) return <Spinner />;
+  useEffect(() => {
+    const worker = new Worker(
+      new URL("../(workers)/impact-curve-calculation.ts", import.meta.url),
+    );
+
+    worker.onmessage = (event: MessageEvent<ImpactWorkerOutputData>) => {
+      const { result, type, swapDirection } = event.data;
+      const key = `${type}Amounts${
+        swapDirection === "in" ? "AnalysisTokenIn" : "TabTokenIn"
+      }`;
+
+      if (key === "initialAmountsAnalysisTokenIn") {
+        setInitialAmounts((prev) => ({ ...prev, analysisTokenIn: result }));
+      } else if (key === "initialAmountsTabTokenIn") {
+        setInitialAmounts((prev) => ({ ...prev, tabTokenIn: result }));
+      } else if (key === "customAmountsAnalysisTokenIn") {
+        setCustomAmounts((prev) => ({ ...prev, analysisTokenIn: result }));
+      } else if (key === "customAmountsTabTokenIn") {
+        setCustomAmounts((prev) => ({ ...prev, tabTokenIn: result }));
+      }
+    };
+
+    worker.postMessage({
+      tokenIn: analysisToken,
+      tokenOut: currentTabToken,
+      data: initialData,
+      poolType: initialData.poolType,
+      swapDirection: "in",
+      type: "initial",
+    } as ImpactWorkerInputData);
 
```

Comment:
this way, you're posting multiple messages to the same worker which will end up making it queue them. You could instead instantiate multiple webworkers to achieve parallel computation given that results don't depend on one another.

Reply:
example here https://github.com/bleu-studio/balancer-tools/pull/209

Reply:
got it! thanks! will do the same for swap curve